### PR TITLE
[CBRD-23490] create online index

### PR DESCRIPTION
**Pull request by razvanradulescu**
http://jira.cubrid.org/browse/CBRD-23490

Create online index.

### CODE REVIEW

File: en/sql/schema/index_stmt.rst [hunk 1]
```diff
@@ -43,6 +46,8 @@ For how to use indexes on the **SELECT** statement like Using SQL Hint, Descendi
     *   The session and server timezone (:ref:`timezone-parameters`) should not be changed if database contains indexes or function index on columns of type TIMESTAMP, TIMESTAMP WITH LOCAL TIME ZONE or DATETIME WITH LOCAL TIME ZONE.
     
     *   The leap second support parameter (:ref:`timezone-parameters`) should not be changed if database contains indexes or function index on columns of type TIMESTAMP or TIMESTAMP WITH LOCAL TIME ZONE. 
+    
+    *   The number of threads from PARALLEL option applies only to index loading step. The heap scan step is performed by the transaction thread (single thread). During this process, the rows are collected into batches; when a batch is full (reached 16M in size), it is dispatched to a index loading thread (or pushed in the queue), the collecting process continues. Batch size computing consists of index key size and the OID size (8 bytes).
```
Comment (ribram) [inline en/sql/schema/index_stmt.rst]: I am afraid this paragraph is too technical (for instance OID will surely confuse any reader).

File: en/sql/schema/index_stmt.rst [hunk 2]
```diff
@@ -87,6 +92,19 @@ Or you can see the index comments with ;sc command in the CSQL interpreter.
...
     csql> ;sc tbl
 
+Online index creation
+---------------------
+
+You can create the index while still allowing other transactions to insert or update the table.
+
+.. code-block:: sql
+
+    CREATE TABLE tbl (a int, b int, c int);
```
Comment (ribram) [inline en/sql/schema/index_stmt.rst]: The first example is not online index loading.

File: en/sql/schema/index_stmt.rst (hunk 1 above)
Comment (ribram) [reply en/sql/schema/index_stmt.rst]: Maybe it is better to offer additional information more relevant to use, like how locking works. It first gets SCH_M lock, then demotes to IX? then it promotes in the end back to SCH_M in both cases successful or failed.

Another information worth mentioning is how unique constraint violation is checked (I think it is reported at the end of loading).

Are there other restrictions or differences to normal loading?

Comment (eseokoh) [reply en/sql/schema/index_stmt.rst]: We may also add an example how csql shows an index being built:
`INDEX idx1 ON usertable (field5) IN PROGRESS`

one more thing also worth addressing is `db_index` catalog table has `status` column. 
online index will be added to the catalog table once it is completed. it won't be shown during online loading. 
you may cover it as a separate patch.

File: en/sql/schema/index_stmt.rst [hunk 3]
```diff
@@ -43,6 +46,20 @@ For how to use indexes on the **SELECT** statement like Using SQL Hint, Descendi
...
+    
+    *   The number of threads from PARALLEL option applies only to index loading step. The heap scan step is performed by the transaction thread (single thread). During this process, the rows are collected into batches; when a batch is full (reached 16M in size), it is dispatched to a index loading thread (or pushed in the queue), the collecting process continues. Batch size computing consists of index key size and the row identifier size (8 bytes).
+
+    *   The creation of online index is performed in three stages:
+    
+        * Adding index into schema with status INDEX IS IN ONLINE BUILDING. This is performed with a SCH_M_LOCK (like all schema changes). After this, the lock is demoted to IX_LOCK.
+
+        * Populating the index : scanning the heap file into batches, sorting the batches and adding the keys to index. During this step, other transactions can modify table data (index is also updated with data changes from other committed transactions).
+
+        * Updating the index status as NORMAL INDEX; this is again performed with a SCH_M_LOCK.
```
Comment (ribram) [inline en/sql/schema/index_stmt.rst]:
```suggestion
        * Updating the index status as NORMAL INDEX; this is performed after promoting the table lock back to SCH_M_LOCK (promotion is guaranteed).
```

File: en/sql/schema/index_stmt.rst [hunk 4]
```diff
@@ -87,6 +104,124 @@ Or you can see the index comments with ;sc command in the CSQL interpreter.
...
+       csql> desc t1;
+       
+       === <Result of SELECT Command in Line 1> ===
+       
+         Field                 Type                  Null                  Key                   Default               Extra
+       ====================================================================================================================================
+         'i1'                  'INTEGER'             'YES'                 'MUL'                 NULL                  ''
+         'i2'                  'INTEGER'             'YES'                 ''                    NULL                  ''
+
+       csql> ;sc t1
```
Comment (ribram) [inline en/sql/schema/index_stmt.rst]:
```suggestion
       csql> ;schema t1
```

File: en/sql/schema/index_stmt.rst [hunk 5]
```diff
@@ -43,6 +46,22 @@ For how to use indexes on the **SELECT** statement like Using SQL Hint, Descendi
     *   The session and server timezone (:ref:`timezone-parameters`) should not be changed if database contains indexes or function index on columns of type TIMESTAMP, TIMESTAMP WITH LOCAL TIME ZONE or DATETIME WITH LOCAL TIME ZONE.
     
     *   The leap second support parameter (:ref:`timezone-parameters`) should not be changed if database contains indexes or function index on columns of type TIMESTAMP or TIMESTAMP WITH LOCAL TIME ZONE. 
+    
+    *   The number of threads from PARALLEL option applies only to index loading step. The heap scan step is performed by the transaction thread (single thread). During this process, the rows are collected into batches; when a batch is full (reached 16M in size), it is dispatched to a index loading thread (or pushed in the queue), the collecting process continues. Batch size computing consists of index key size and the row identifier size (8 bytes).
+
+    *   When using CSQL in stand-alone mode, the WITH ONLINE option is ignored (normal index is created).
```
Comment (eseokoh) [inline en/sql/schema/index_stmt.rst]:
```suggestion
    *   Under stand-alone mode, the WITH ONLINE option is ignored (normal index is created).
```

File: en/sql/schema/index_stmt.rst [hunk 6]
```diff
@@ -43,6 +46,22 @@ For how to use indexes on the **SELECT** statement like Using SQL Hint, Descendi
...
+
+    *   The creation of online index is performed in three stages:
+    
+        * Adding index into schema with status INDEX IS IN ONLINE BUILDING. This is performed with a SCH_M_LOCK (like all schema changes) on table. After this, the lock is demoted to IX_LOCK.
+
+        * Populating the index : scanning the heap file into batches, sorting the batches and adding the keys to index. During this step, other transactions can modify table data (index is also updated with data changes from other committed transactions).
+
+        * Updating the index status as NORMAL INDEX; this is performed after promoting the table lock back to SCH_M_LOCK (promotion is guaranteed).
+    
+    *   The online index being build is displayed by SHOW statements from other transactions. It is not visible from other transactions in :ref:`db-index` system table due to MVCC snapshot (other transactions can only see committed entries in this table).
```
Comment (eseokoh) [inline en/sql/schema/index_stmt.rst]:
```suggestion
    *   The online index being built is displayed by SHOW statements from other transactions. It is not visible from other transactions in :ref:`db-index` system table due to MVCC snapshot (other transactions can only see committed entries in this table).
```

**Pull request merged**